certora/specs/Delegation.spec

```diff
@@ -0,0 +1,103 @@
+import "ERC20.spec";
+
+methods {
+    function _moveDelegateVotesExternal(address from, address to, uint256 amount) external envfree;
+    function _burnExternal(address, uint256) external envfree;
+    function ERC20Upgradeable._burn(address account, uint256 value) internal
+        => summaryBurn(account, value);
+    function DelegationToken._moveDelegateVotes(address from, address to, uint256 amount) internal
+        => summaryMove(from, to, amount);
+}
+
+ghost mathint sumOfVotingPower {
+    init_state axiom sumOfVotingPower == 0;
+}
+
+// The ghost variable burnCalled is used to track calls to burn.
+persistent ghost bool burnCalled {
+    init_state axiom burnCalled == false;
+}
+
+function summaryBurn(address a, uint256 amount) {
+    // Mark _burn as being called.
+    burnCalled = true;
+
+    // Burn tokens.
+    _burnExternal(a, amount);
+}
+
+function summaryMove(address from, address to, uint256 amount) {
+    if (burnCalled) {
+        // Ensure that burnt voting power is deducted from sumOfVotingPower.
+        sumOfVotingPower = sumOfVotingPower - amount;
+        burnCalled = false;
+    } else if (from == 0 && to != 0) {
+        // Ensure that voting power is deducted from sumOfVotingPower during a transfer or a delegation when delegatee is zero address.
+        sumOfVotingPower = sumOfVotingPower - amount;
```

Comment:
This misses minting I think with that substraction, when minting, the `sumOfVotingPower` would stay the same. This makes the verification pass, because we are checking an inequality. I think it highlights why it would be best to have the simplest ghost variables possible, to have a clear view of what they represent, and then we can have more complex specs

Reply:
You're right, for instance if we strengthen the invariant by checking that sum of total votes is at least zero, it will fail. This also why I like better the solution whit a ghost variable.

Reply:
Fixed by https://github.com/morpho-org/morpho-token-upgradeable/pull/85/commits/769c78651abf35c2b697342c8746d122c740bc7a.